File: client/python/angzarr_client/saga_context.py
"""SagaContext for splitter pattern support.

SagaContext provides convenient access to destination aggregate state
in the splitter pattern, where one event triggers commands to multiple aggregates.

Example usage:
    def handle_table_settled(event: TableSettled, destinations: list[EventBook]) -> list[CommandBook]:
        ctx = SagaContext(destinations)
        commands = []

        for payout in event.payouts:
            seq = ctx.get_sequence("player", payout.player_root)
            cmd = TransferFunds(player_root=payout.player_root, amount=payout.amount)
            commands.append(new_command_book("player", cmd, sequence=seq))

        return commands
"""

from __future__ import annotations

from .helpers import next_sequence, root_id_hex
from .proto.angzarr import types_pb2 as types

__all__ = ["SagaContext"]


class SagaContext:
    """Context for saga handlers providing access to destination aggregate state.

    Used in the splitter pattern where one event triggers commands to multiple
    aggregates. Provides sequence number lookup for optimistic concurrency control.
    """

    def __init__(self, destinations: list[types.EventBook]) -> None:
        """Create a context from a list of destination EventBooks.

        Args:
            destinations: List of EventBooks fetched during prepare phase.
        """
        self._destinations: dict[str, types.EventBook] = {}
        for book in destinations:
            if book.HasField("cover") and book.cover.domain:
                key = self._make_key(book.cover.domain, book.cover.root.value)
                self._destinations[key] = book

    def get_sequence(self, domain: str, aggregate_root: bytes) -> int:
        """Get the next sequence number for a destination aggregate.

        Returns 1 if the aggregate doesn't exist yet.

        Args:
            domain: The domain of the target aggregate.
            aggregate_root: The root identifier (as bytes).

        Returns:
            The next sequence number for the aggregate.
        """
        key = self._make_key(domain, aggregate_root)
        book = self._destinations.get(key)
        if book is None:
            return 1
        return next_sequence(book)

    def get_destination(
        self, domain: str, aggregate_root: bytes
    ) -> types.EventBook | None:
        """Get the EventBook for a destination aggregate.

        Args:
            domain: The domain of the target aggregate.
            aggregate_root: The root identifier (as bytes).

        Returns:
            The EventBook if found, None otherwise.
        """
        key = self._make_key(domain, aggregate_root)
        return self._destinations.get(key)

    def has_destination(self, domain: str, aggregate_root: bytes) -> bool:
        """Check if a destination exists.

        Args:
            domain: The domain of the target aggregate.
            aggregate_root: The root identifier (as bytes).

        Returns:
            True if the destination exists.
        """
        key = self._make_key(domain, aggregate_root)
        return key in self._destinations

    @staticmethod
    def _make_key(domain: str, root: bytes) -> str:
        """Create a lookup key from domain and root."""
        return f"{domain}:{root.hex()}"

File: client/csharp/Angzarr.Client/SagaContext.cs
using Google.Protobuf;

namespace Angzarr.Client;

/// <summary>
/// Context for saga handlers, providing access to destination aggregate state.
///
/// Used in the splitter pattern where one event triggers commands to multiple aggregates.
/// Provides sequence number lookup for optimistic concurrency control.
///
/// <code>
/// List&lt;CommandBook&gt; HandleTableSettled(TableSettled evt, SagaContext ctx)
/// {
///     var commands = new List&lt;CommandBook&gt;();
///     foreach (var payout in evt.Payouts)
///     {
///         var seq = ctx.GetSequence("player", payout.PlayerRoot);
///         var cmd = new TransferFunds { PlayerRoot = payout.PlayerRoot, Amount = payout.Amount };
///         commands.Add(NewCommandBook("player", cmd, seq));
///     }
///     return commands;
/// }
/// </code>
/// </summary>
public class SagaContext
{
    private readonly Dictionary<string, Angzarr.EventBook> _destinations;

    /// <summary>
    /// Create a context from a list of destination EventBooks.
    /// </summary>
    /// <param name="destinationBooks">List of EventBooks fetched during prepare phase.</param>
    public SagaContext(IEnumerable<Angzarr.EventBook> destinationBooks)
    {
        _destinations = new Dictionary<string, Angzarr.EventBook>();
        foreach (var book in destinationBooks)
        {
            if (book.Cover != null && !string.IsNullOrEmpty(book.Cover.Domain))
            {
                var key = MakeKey(book.Cover.Domain, book.Cover.Root?.Value ?? ByteString.Empty);
                _destinations[key] = book;
            }
        }
    }

    /// <summary>
    /// Get the next sequence number for a destination aggregate.
    /// Returns 1 if the aggregate doesn't exist yet.
    /// </summary>
    /// <param name="domain">The domain of the target aggregate.</param>
    /// <param name="aggregateRoot">The root identifier.</param>
    /// <returns>The next sequence number for the aggregate.</returns>
    public uint GetSequence(string domain, ByteString aggregateRoot)
    {
        var key = MakeKey(domain, aggregateRoot);
        if (_destinations.TryGetValue(key, out var book))
        {
            if (book.Pages.Count == 0)
                return 1;
            var lastPage = book.Pages[^1];
            return (lastPage.Header?.Sequence ?? 0) + 1;
        }
        return 1;
    }

    /// <summary>
    /// Get the next sequence number for a destination aggregate.
    /// Returns 1 if the aggregate doesn't exist yet.
    /// </summary>
    /// <param name="domain">The domain of the target aggregate.</param>
    /// <param name="aggregateRoot">The root identifier as bytes.</param>
    /// <returns>The next sequence number for the aggregate.</returns>
    public uint GetSequence(string domain, byte[] aggregateRoot)
    {
        return GetSequence(domain, ByteString.CopyFrom(aggregateRoot));
    }

    /// <summary>
    /// Get the EventBook for a destination aggregate, if available.
    /// </summary>
    /// <param name="domain">The domain of the target aggregate.</param>
    /// <param name="aggregateRoot">The root identifier.</param>
    /// <returns>The EventBook if found, null otherwise.</returns>
    public Angzarr.EventBook? GetDestination(string domain, ByteString aggregateRoot)
    {
        var key = MakeKey(domain, aggregateRoot);
        return _destinations.TryGetValue(key, out var book) ? book : null;
    }

    /// <summary>
    /// Get the EventBook for a destination aggregate, if available.
    /// </summary>
    /// <param name="domain">The domain of the target aggregate.</param>
    /// <param name="aggregateRoot">The root identifier as bytes.</param>
    /// <returns>The EventBook if found, null otherwise.</returns>
    public Angzarr.EventBook? GetDestination(string domain, byte[] aggregateRoot)
    {
        return GetDestination(domain, ByteString.CopyFrom(aggregateRoot));
    }

    /// <summary>
    /// Check if a destination exists.
    /// </summary>
    /// <param name="domain">The domain of the target aggregate.</param>
    /// <param name="aggregateRoot">The root identifier.</param>
    /// <returns>True if the destination exists.</returns>
    public bool HasDestination(string domain, ByteString aggregateRoot)
    {
        var key = MakeKey(domain, aggregateRoot);
        return _destinations.ContainsKey(key);
    }

    private static string MakeKey(string domain, ByteString root)
    {
        return $"{domain}:{root.ToBase64()}";
    }
}
